engine/src/flutter/lib/ui/painting/pixel_deferred_image_gpu_impeller.cc

```diff
@@ -0,0 +1,133 @@
+    fml::RefPtr<fml::TaskRunner> raster_task_runner) {
+  return sk_sp<PixelDeferredImageGPUImpeller>(new PixelDeferredImageGPUImpeller(
+      PixelDeferredImageGPUImpeller::ImageWrapper::Make(
+          std::move(image), std::move(snapshot_delegate),
+          std::move(raster_task_runner))));
+}
+
+PixelDeferredImageGPUImpeller::PixelDeferredImageGPUImpeller(
+    std::shared_ptr<ImageWrapper> wrapper)
+    : wrapper_(std::move(wrapper)) {}
+
+PixelDeferredImageGPUImpeller::~PixelDeferredImageGPUImpeller() = default;
+
+sk_sp<SkImage> PixelDeferredImageGPUImpeller::skia_image() const {
+  return nullptr;
+}
+
+std::shared_ptr<impeller::Texture>
+PixelDeferredImageGPUImpeller::impeller_texture() const {
+  if (!wrapper_) {
+    return nullptr;
+  }
+  return wrapper_->texture();
+}
+
+bool PixelDeferredImageGPUImpeller::isOpaque() const {
+  return false;
+}
+
+bool PixelDeferredImageGPUImpeller::isTextureBacked() const {
+  return wrapper_ && wrapper_->isTextureBacked();
+}
+
+bool PixelDeferredImageGPUImpeller::isUIThreadSafe() const {
+  return true;
+}
+
+DlISize PixelDeferredImageGPUImpeller::GetSize() const {
+  return wrapper_ ? wrapper_->size() : DlISize();
+}
+
+size_t PixelDeferredImageGPUImpeller::GetApproximateByteSize() const {
+  auto size = sizeof(PixelDeferredImageGPUImpeller);
+  if (wrapper_) {
+    if (wrapper_->texture()) {
+      size += wrapper_->texture()
+                  ->GetTextureDescriptor()
+                  .GetByteSizeOfBaseMipLevel();
+    } else {
+      size += wrapper_->size().Area() * 4;
+    }
+  }
+  return size;
+}
+
+std::optional<std::string> PixelDeferredImageGPUImpeller::get_error() const {
+  return wrapper_ ? wrapper_->get_error() : std::nullopt;
+}
+
+std::shared_ptr<PixelDeferredImageGPUImpeller::ImageWrapper>
+PixelDeferredImageGPUImpeller::ImageWrapper::Make(
+    sk_sp<SkImage> image,
+    fml::TaskRunnerAffineWeakPtr<SnapshotDelegate> snapshot_delegate,
+    fml::RefPtr<fml::TaskRunner> raster_task_runner) {
+  auto wrapper = std::shared_ptr<ImageWrapper>(new ImageWrapper(
+      image, std::move(snapshot_delegate), std::move(raster_task_runner)));
+  wrapper->SnapshotImage(std::move(image));
+  return wrapper;
+}
+
+PixelDeferredImageGPUImpeller::ImageWrapper::ImageWrapper(
+    const sk_sp<SkImage>& image,
+    fml::TaskRunnerAffineWeakPtr<SnapshotDelegate> snapshot_delegate,
+    fml::RefPtr<fml::TaskRunner> raster_task_runner)
+    : size_(DlISize(image->width(), image->height())),
+      snapshot_delegate_(std::move(snapshot_delegate)),
+      raster_task_runner_(std::move(raster_task_runner)) {}
+
+PixelDeferredImageGPUImpeller::ImageWrapper::~ImageWrapper() = default;
+
+bool PixelDeferredImageGPUImpeller::ImageWrapper::isTextureBacked() const {
+  return texture_ && texture_->IsValid();
+}
+
+void PixelDeferredImageGPUImpeller::ImageWrapper::SnapshotImage(
+    sk_sp<SkImage> image) {
+  fml::TaskRunner::RunNowOrPostTask(
```

Comment:
This may only be called once right? Perhaps guard against incorrect use?

Reply:
done